pkg/tournament/tournament.go

```diff
@@ -550,30 +788,95 @@ func SetResult(ctx context.Context,
 		return err
 	}
 
+	err = possiblyEndTournament(ctx, ts, t, division)
+	if err != nil {
+		return err
+	}
+
 	err = ts.Set(ctx, t)
 	if err != nil {
 		return err
 	}
-	return SendTournamentDivisionMessage(ctx, ts, id, division)
+	pairingsMessage := PairingsToResponse(id, division, pairing, standings)
+	wrapped := entity.WrapEvent(pairingsMessage, realtime.MessageType_TOURNAMENT_DIVISION_PAIRINGS_MESSAGE)
+	return SendTournamentMessage(ctx, ts, id, wrapped)
+}
+
+func possiblyEndTournament(ctx context.Context, ts TournamentStore, t *entity.Tournament,
+	division string) error {
+
+	divisionObject := t.Divisions[division]
+	ended, err := divisionObject.DivisionManager.IsFinished()
+	if err != nil {
+		return err
+	}
+	allended := ended
+	if ended {
+		for dname, div := range t.Divisions {
+			if dname == division {
+				continue
+				// no need to check again
+			}
+			dended, err := div.DivisionManager.IsFinished()
+			if err != nil {
+				return err
+			}
+			if !dended {
+				allended = false
+				break
+			}
+		}
+	}
+	if allended {
+		t.IsFinished = true
+		err := ts.RemoveRegistrantsForTournament(ctx, t.UUID)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
 }
 
-func StartTournament(ctx context.Context, ts TournamentStore, id string, manual bool) error {
+func StartAllRoundCountdowns(ctx context.Context, ts TournamentStore, id string, round int) error {
 	t, err := ts.Get(ctx, id)
 	if err != nil {
 		return err
 	}
 
 	// Do not lock, StartRound will do that
 
+	if t.IsFinished {
+		return fmt.Errorf("tournament %s has finished", id)
+	}
+
+	if len(t.Divisions) == 0 {
+		return fmt.Errorf("cannot start tournament %s with no divisions", t.Name)
+	}
+
 	for division := range t.Divisions {
-		t.IsStarted = true
-		err := StartRoundCountdown(ctx, ts, id, division, 0, manual, false)
+		dm := t.Divisions[division].DivisionManager
+		if dm == nil {
+			return fmt.Errorf("cannot start round %d for division %s because it has a nil division manager", round, division)
+		}
+		if dm.GetDivisionControls().GameRequest == nil {
+			return fmt.Errorf("no division game controls have been set for division %v", division)
+		}
+		isReady, err := dm.IsRoundReady(round)
 		if err != nil {
 			return err
 		}
+		if !isReady {
+			return fmt.Errorf("cannot start round %d for division %s because the round is not ready", round, division)
+		}
 	}
-	if !t.IsStarted {
-		return fmt.Errorf("cannot start tournament %s with no divisions", t.Name)
+
+	for division := range t.Divisions {
+		t.IsStarted = true
+		err := StartRoundCountdown(ctx, ts, id, division, round, false, true)
```

Comment:
just noting here
- `StartAllRoundCountdowns` calls `ts.Get` but does not lock
https://github.com/domino14/liwords/blob/5af38e6d1aa446e15bdeb1a2aca40e0ce981add8/pkg/tournament/tournament.go#L842-L848
- `StartAllRoundCooldowns` calls `StartRoundCountdown` (here) for each division
- `StartRoundCooldown` calls `ts.Get` again and locks/unlocks each time
https://github.com/domino14/liwords/blob/5af38e6d1aa446e15bdeb1a2aca40e0ce981add8/pkg/tournament/tournament.go#L893-L901
- but `ts.Get` is not obligated to return the same object
- it's not immediately clear that this `StartAllRoundCooldowns` depends on `ts` implementing a caching mechanism

@jvc56